src/modules/schnorrsig/main_impl.h

```diff
@@ -0,0 +1,236 @@
+    unsigned char seckey[32];
+    int ret = 1;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    ARG_CHECK(sig64 != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(keypair != NULL);
+
+    if (noncefp == NULL) {
+        noncefp = secp256k1_nonce_function_bip340;
+    }
+
+    ret &= secp256k1_keypair_load(ctx, &sk, &pk, keypair);
+    /* Because we are signing for a x-only pubkey, the secret key is negated
+     * before signing if the point corresponding to the secret key does not
+     * have an even Y. */
+    secp256k1_fe_normalize(&pk.y);
+    if (secp256k1_fe_is_odd(&pk.y)) {
+        secp256k1_scalar_negate(&sk, &sk);
+    }
+
+    secp256k1_scalar_get_b32(seckey, &sk);
+    secp256k1_fe_normalize(&pk.x);
+    secp256k1_fe_get_b32(pk_buf, &pk.x);
+    ret &= !!noncefp(buf, msg32, seckey, pk_buf, bip340_algo16, (void*)ndata, 0);
+    memset(seckey, 0, sizeof(seckey));
+    secp256k1_scalar_set_b32(&k, buf, NULL);
+    ret &= !secp256k1_scalar_is_zero(&k);
+    secp256k1_scalar_cmov(&k, &secp256k1_scalar_one, !ret);
+
+    secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &rj, &k);
+    secp256k1_ge_set_gej(&r, &rj);
+
+    /* We declassify r to allow using it as a branch point. This is fine
+     * because r is not a secret. */
+    secp256k1_declassify(ctx, &r, sizeof(r));
+    if (!secp256k1_fe_is_quad_var(&r.y)) {
+        secp256k1_scalar_negate(&k, &k);
+    }
+    secp256k1_fe_normalize(&r.x);
+    secp256k1_fe_get_b32(&sig64[0], &r.x);
+
+    /* tagged hash(r.x, pk.x, msg32) */
+    secp256k1_schnorrsig_sha256_tagged(&sha);
+    secp256k1_sha256_write(&sha, &sig64[0], 32);
+    secp256k1_sha256_write(&sha, pk_buf, sizeof(pk_buf));
+    secp256k1_sha256_write(&sha, msg32, 32);
+    secp256k1_sha256_finalize(&sha, buf);
+
+    /* Set scalar e to the challenge hash modulo the curve order as per
+     * BIP340. */
+    secp256k1_scalar_set_b32(&e, buf, NULL);
+    secp256k1_scalar_mul(&e, &e, &sk);
+    secp256k1_scalar_add(&e, &e, &k);
+    secp256k1_scalar_get_b32(&sig64[32], &e);
+
+    memczero(sig64, 64, !ret);
+    secp256k1_scalar_clear(&k);
+    secp256k1_scalar_clear(&sk);
+
+    return ret;
+}
+
+int secp256k1_schnorrsig_verify(const secp256k1_context* ctx, const unsigned char *sig64, const unsigned char *msg32, const secp256k1_xonly_pubkey *pubkey) {
+    secp256k1_scalar s;
+    secp256k1_scalar e;
+    secp256k1_gej rj;
+    secp256k1_ge pk;
+    secp256k1_gej pkj;
+    secp256k1_fe rx;
+    secp256k1_sha256 sha;
+    unsigned char buf[32];
+    int overflow;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_context_is_built(&ctx->ecmult_ctx));
+    ARG_CHECK(sig64 != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(pubkey != NULL);
+
+    if (!secp256k1_fe_set_b32(&rx, &sig64[0])) {
+        return 0;
+    }
+
+    secp256k1_scalar_set_b32(&s, &sig64[32], &overflow);
+    if (overflow) {
+        return 0;
+    }
```

Comment:
The BIP says that these can't be 0 either, (`fail if r ≥ p` and ` fail if s ≥ n`)
having `rx = 0` won't help an attacker because you make sure that `rj` isn't infinity when returning.
but having `s = 0` allows you to get `rj = -e*pkj` but then you have the problem you need to solve `-e*pkj` to get your `r` but you need the `r` to produce `e`, so this relies on the properties of the hash function

Reply:
The BIP allows s = 0 and r = 0. Also solving `rj = -e*pkj` isn't easier than `rj - s*G = -e*pkj` without the secret key or am I misunderstanding the attack scenario?

Reply:
it is, because in `rj - s*G = -e*pkj`, you want to find the *scalar* `s`, which is DLP.
but in `rj = -e*pkj` you want to find `rj` which is a point. this isn't DLP, it's still "hard" though because `rj` should be hashed inside of `e`.

about the bip, doesn't this say that zeroes should fail? (if `r=p` or `s=n`)
```
Let r = int(sig[0:32]); fail if r ≥ p.
Let s = int(sig[32:64]); fail if s ≥ n.
```

(I have deja vu about this discussion but I can't find it in the review history hehe )

Reply:
I don't see how forging is supposed to be easier for s = 0 than for s != 0. If the hash function is "broken", you can forge without solving DLP but you don't need s = 0 for this. Fix any s and solve sG=R+H(P,R,m)P for R. 


> about the bip, doesn't this say that zeroes should fail? (if `r=p` or `s=n`)

It holds that `s = int(0x00...00) = 0 <= n`.

Reply:
Ha, you're right, I don't know how I missed that.

>It holds that `s = int(0x00...00) = 0 <= n`.

so you're saying that `0 <= n`? even though it's a scalar and it's mod n (so `0 mod n == n mod n`)

Reply:
`s` is an unsigned integer here in the pseudocode.

"The function int(x), where x is a 32-byte array, returns the 256-bit unsigned integer whose most significant byte first encoding is x."

Reply:
> I don't see how forging is supposed to be easier for s = 0 than for s != 0. If the hash function is "broken", you can forge without solving DLP but you don't need s = 0 for this. Fix any s and solve sG=R+H(P,R,m)P for R.

Okay strictly speaking, one can think of a failure mode of the hash function where this is easier to solve for s = 0, but then I don't see that this failure mode is more likely than any corresponding failure mode some other fixed s != 0.

Reply:
> > I don't see how forging is supposed to be easier for s = 0 than for s != 0. If the hash function is "broken", you can forge without solving DLP but you don't need s = 0 for this. Fix any s and solve sG=R+H(P,R,m)P for R.
> 
> Okay strictly speaking, one can think of a failure mode of the hash function where this is easier to solve for s = 0, but then I don't see that this failure mode is more likely than any corresponding failure mode some other fixed s != 0.

Actually, strictly speaking, one can't! As it still has not been garbage collected from when I was looking at Taproot hash function requirements, allow me to recount the formal hash function requirements for *key-only* Schnorr unforgeability from Neven et al.: The adversary must be unable to break the *random prefix-preimage* (rpp) security of the hash. This is necessary and sufficient in the generic group model. In the rpp game, the adversary is allowed to chose the image `c` and then is given a random challenge prefix `R`. They win the game if they can output `m` such that `H(R||m) == c`.

To see that rpp is necessary and that there is no value for `s` for which the scheme is safer against a hash break, consider that with a rrp solver `A` we can forge a signature with the following algorithm (ignoring key-prefixing for simplicity):
```
Forge(X):
1. c <--$ A
2. choose any s you want
3. R <-- sG + cX
4. m <--$ A(R)
5. return (R,s),m
```
Thus if the hash function breaks in any way that makes signatures forgeable then rpp must be broken as well (as it is sufficient) therefore an efficient `A`  exists and they will be forgeable for any choice of `s`. They are not necessarily easily forgeable for any value of `c`. Perhaps if finding a pre-image for 0 was easier than other values then banning `c=0` would help (but it would be better to just stop using the hash function in this case).
arbitrary
Original argument that I have hopefully recalled correctly:  http://www.neven.org/papers/schnorr.pdf

Reply:
Thanks, that's super interesting.